### ISSUE TITLE
ComboBox inconsistently clearing text when not matched in listbox

### ISSUE THREAD

**Issue by tznind** [gui-cs/Terminal.Gui#1704]
**Describe the bug**
There is a wierd bug with ComboBox where typing text into the box that does not appear in its Source gets cleared when you leave the control.  The issue goes away if you open then collapse the listbox first.

**To Reproduce**
1. Run ComboBoxIteration Scenario
2. Type into the ComboBox "hey there"
3. Press tab
4. See error: text is cleared

Workaround:
1. Run ComboBoxIteration Scenario
2. Expand then collapse the ComboBox manually with mouse
3. Type into the ComboBox "hey there"
4. Press tab
5. Now the text is not cleared

**Expected behavior**
Text should not be cleared upon leaving the combo box.

**Screenshots**
![ComboBoxClearing](https://user-images.githubusercontent.com/31306100/167584915-9703cfb4-3107-4073-970f-367e9f5d7626.gif)

**Desktop (please complete the following information):**
 - OS: Windows
 - Shell: Powershell

**Comment by BDisp** [comment]
There is many situations where we want a read only combo box and in this cases must not allowing adding more items to the source.

**Comment by tznind** [comment]
At the moment though the behaviour is inconsistent.  

If you toggle the dropdown with the mouse then it doesn't force the Text to match the list.  While if you don't toggle the list it does force the Text to match.

**Comment by BDisp** [comment]
Yes, I know it is inconsistent and should be corrected. I just wanted to draw attention to this detail. The author of this control just intended to create a search box when typing in `TextField`, such as the `AutoComplete`, and I proposed that it could be transformed into a `Combobox`, so it is still inconsistent. Therefore, any improvement that can be done only brings benefits.


> I guess there are 2 questions here:
> 
>     1. Is it supposed to clear it or not?
> 
If it's a read only and the typed text does not matching the source then should clear. If match then it must show the equivalents.
If not read only then the text typed can stay in the `TextField` and if needed to add to the source then must have an action by the developer to do the addition.

>     2. Why does not not clear it if you toggle the dropdown down/up with mouse first
>
Must have the same behavior as 1.

**Comment by BDisp** [comment]
> How do you control if a ComboBox is 'read only'. Or are you talking about the way we want it to be rather than how it currently is?

That is. There is a `ReadOnly` property on the `ComboBox` that don't allow typing, I think, but the user may want to type to search for it. It's complicated :-)

**Comment by tznind** [comment]
Ok cool.  Thanks for explaining.  

In the interests of documenting the current system state.  My current workaround to force the combo box not to clear search is just to toggle the box open/collapse manually in code.

```csharp
var expand = typeof(ComboBox).GetMethod("Expand");
var collapse = typeof(ComboBox).GetMethod("Collapse");
expand.Invoke(combo, new object[0]);
collapse.Invoke(combo,new object[0]);
```

For personal preference I also like setting up this keybinding which makes the interaction feel more natural:

```csharp
combo.AddKeyBinding(Key.CursorDown, Command.Expand);
```

**Comment by BDisp** [comment]
> For personal preference I also like setting up this keybinding which makes the interaction feel more natural:
> 
> ```cs
> combo.AddKeyBinding(Key.CursorDown, Command.Expand);
> ```

Does this not will conflict with the` AddKeyBinding (Key.CursorDown, Command.LineDown);`?

**Comment by tznind** [comment]
> Does this not will conflict with the` AddKeyBinding (Key.CursorDown, Command.LineDown);`?

It doesn't seem to.  New keybindings overwrite old ones so theres no Exception and I think because there focus changes on Expand to the listbox it all just seems to work (i.e. it doesn't break up/down).

**Comment by BDisp** [comment]
> It doesn't seem to. New keybindings overwrite old ones so theres no Exception. And I think because the input focus changes on Expand to the listbox it all just seems to work (i.e. it doesn't break up/down).

Ah right, it will change the focus to the `ListView` and are this view that will handle it, thanks for the clarification.

**Comment by BDisp** [comment]
@tznind, @tig I think the `ComboBox` should always clear if the typed text doesn't exist on the source.

![combobox-clear](https://user-images.githubusercontent.com/13117724/226930646-4baa12e2-ef71-4c84-be1f-daf47153285c.gif)

If the `ComboBox` is open by F4 or mouse the `ListView` will always be focused and the first item or the selected item will be the current, updating the search text through the `SelectedItemChanged` event when cursor keys are used. Pressing `Key.Escape` undoes the changes and close the `ListView`.

![combobox-open](https://user-images.githubusercontent.com/13117724/226932606-205b125a-101e-4913-b262-7999ff038e9a.gif)

Typing the exactly text or navigation through cursor to the item, pressing `Key.Enter` or tabbing will accept the change.

![combobox-accept-leave](https://user-images.githubusercontent.com/13117724/226944692-b6ade12a-3b4a-4733-b056-d5d33c9c95a0.gif)

If `ReadOnly` is enabled only the first letter will be used on search.

![combobox-readonly](https://user-images.githubusercontent.com/13117724/226959449-319563b5-7704-4658-aa3a-737b0e19a1bb.gif)

This is the iteration with a already selected item:

![combobox-iteration](https://user-images.githubusercontent.com/13117724/226962872-b0201d5d-8e80-41a3-91a9-75427eea2222.gif)

This is a `ComboBox` alone in a `Dialog` as the author initially created and it never isn't auto hide. I want to ask if is this to maintain or not?

![combobox-dialog](https://user-images.githubusercontent.com/13117724/226977376-a13d2af3-f76b-4355-ab8e-9db37e74e005.gif)

The `ListView` shouldn't be a subview of `ComboBox` , but be subview of Application.Top to allow the drop-down always visible even if not fit the superview of the `ComboBox` . Here a reason why it should:

![combobox-window](https://user-images.githubusercontent.com/13117724/226979213-640b6501-b44a-433a-b37f-730e71c7be27.gif)

I'm still working on this but i need answers to my questions above, please. I think the `ComboBox` alone in a `Dialog` can be dropped, because it will be unused, I think.

**Comment by BDisp** [comment]
@tznind, @tig which ColorScheme do you prefer, 1 or 2?
1:
![imagem](https://user-images.githubusercontent.com/13117724/227658813-afa15a71-3723-4740-bc38-07950ff0e6b5.png)
2:
![imagem](https://user-images.githubusercontent.com/13117724/227658899-42043730-ac64-4839-af29-220c3d42ed2a.png)

I vote for 1.

**Comment by tznind** [comment]
1 does have more visual distinction and would look better on Dialogs where the background is grey by default.  So yeah I agree 1.

**Comment by tznind** [comment]
> @tznind, @tig I think the ComboBox should always clear if the typed text doesn't exist on the source.

I think this needs a setting e.g. [DropDownStyle](https://learn.microsoft.com/en-us/dotnet/api/system.windows.forms.combobox.dropdownstyle?view=windowsdesktop-7.0) in WinForms:

DropDown | 1 | Specifies that the list is displayed by clicking the down  arrow and that the text portion is editable. This means that the user  can enter a new value and is not limited to selecting an existing value  in the list. When using this setting, the Append value of AutoCompleteMode works the same as the SuggestAppend value. This is the default style.
-- | -- | --
DropDownList | 2 | Specifies that the list is displayed by clicking the down  arrow and that the text portion is not editable. This means that the  user cannot enter a new value. Only values already in the list can be  selected. The list displays only if AutoCompleteMode is Suggest or SuggestAppend.

There are many cases where you want the user to be able to enter their own value e.g.

![image](https://user-images.githubusercontent.com/31306100/227707604-50885f4e-3e70-4ab5-9565-f5fec72e8ba3.png)
_User can enter any link they want_

![image](https://user-images.githubusercontent.com/31306100/227707745-b4e065ba-c1e9-4cea-b4c2-613cd164f59b.png)
_User could enter 7.5pt if they wanted_

But likewise there are cases where you don't want the user to be able to select any option:

![image](https://user-images.githubusercontent.com/31306100/227707829-21ecad0d-919c-4f2b-bac2-3ffc918916ab.png)
_User can only pick from the selection_

> I'm still working on this but i need answers to my questions above, please. I think the `ComboBox` alone in a `Dialog` can be dropped, because it will be unused, I think.

**Comment by BDisp** [comment]
>>    I'm still working on this but i need answers to my questions above, please. I think the ComboBox alone in a Dialog can be dropped, because it will be unused, I think.

> Thanks for looking into this! I know we can make an awesome view here.

@tznind thanks for your reply. I'm really thinking about replacing the `ComboBox alone in a Dialog` with the `Simple` of the `ComboBoxStyle`:

<html><body>


Simple | 0 | Specifies that the list is always visible and that the text  portion is editable. This means that the user can enter a new value and  is not limited to selecting an existing value in the list.
-- | -- | --



</body>
</html>

In this case the drop down list will be a subview of the `ComboBox` and will always visible and it must positioned in a way that it must fit the `ComboBox` dimensions, like the `Dialog`, but without it. What do you think?

**Comment by tig** [comment]
Closing as ComboBox is going to be replaced asap.
